### PR TITLE
Access stored messages by URL

### PR DESCRIPTION
**Pull request by pirogoeth**
Addresses issues in #168

### CODE REVIEW

File: src/Mailgun/Mailgun.php [hunk 1]
```diff
@@ -36,20 +36,83 @@ class Mailgun
     protected $apiKey;
 
     /**
+     * @var [string => string]
+     */
+    protected $regions = [
```
Comment (Nyholm) [inline src/Mailgun/Mailgun.php]: This should be private.

File: src/Mailgun/Mailgun.php [hunk 2]
```diff
@@ -36,20 +36,83 @@ class Mailgun
...
+        'si' => 'si.api.mailgun.net',
+        'so' => 'so.api.mailgun.net',
+    ];
+
+    /**
+     * @var [string => RestClient]
+     */
+    protected $regionClients = [];
```
Comment (Nyholm) [inline src/Mailgun/Mailgun.php]: This should be private.

File: src/Mailgun/Mailgun.php [hunk 3]
```diff
@@ -36,20 +36,83 @@ class Mailgun
...
     /**
+     * @var [string => string]
+     */
+    protected $regions = [
+        'si' => 'si.api.mailgun.net',
+        'so' => 'so.api.mailgun.net',
+    ];
+
+    /**
+     * @var [string => RestClient]
```
Comment (Nyholm) [inline src/Mailgun/Mailgun.php]: `@var array [string => RestClient]`

Right? Or else PHPDoc will not pick this up. Same with $regions.

File: src/Mailgun/Mailgun.php [hunk 4]
```diff
@@ -36,20 +36,83 @@ class Mailgun
...
      * @param HttpClient  $httpClient
      * @param string      $apiEndpoint
+     * @param bool        $noRegions
      */
     public function __construct(
         $apiKey = null,
         HttpClient $httpClient = null,
-        $apiEndpoint = 'api.mailgun.net'
+        $apiEndpoint = 'api.mailgun.net',
+        bool $noRegions = false
```
Comment (Nyholm) [inline src/Mailgun/Mailgun.php]: We are not exclusive for PHP7. Can't have `bool`.

It is generally bad practice to have feature flags like this.

File: src/Mailgun/Mailgun.php [hunk 5]
```diff
@@ -36,20 +36,83 @@ class Mailgun
...
     ) {
         $this->apiKey = $apiKey;
         $this->restClient = new RestClient($apiKey, $apiEndpoint, $httpClient);
+
+        if ($noRegions !== true) {
+            $this->initializeRegionClients($httpClient);
```
Comment (Nyholm) [inline src/Mailgun/Mailgun.php]: This constructor would create 3 `RestClient` objects, right?

File: src/Mailgun/Mailgun.php [hunk 6]
```diff
@@ -36,20 +36,83 @@ class Mailgun
...
+    {
+        foreach ($this->regions as $region => $apiUrl) {
+            if (strpos($url, 'https://' . $apiUrl) === 0) {
+                return $region;
+            }
+        }
     }
 
     /**
+     * Returns the client for the specified region.
```
Comment (Nyholm) [inline src/Mailgun/Mailgun.php]: This seams to be a loose comment.

File: src/Mailgun/Mailgun.php [hunk 7]
```diff
@@ -208,4 +279,28 @@ public function BatchMessage($workingDomain, $autoSend = true)
     {
         return new BatchMessage($this->restClient, $workingDomain, $autoSend);
     }
+
+    /**
+     * @param string $messageUrl
+     *
+     * @return \stdClass
+     */
+    public function getMessageFromUrl($messageUrl = null)
```
Comment (Nyholm) [inline src/Mailgun/Mailgun.php]: Why would we allow this function to be called with no arguments?

File: src/Mailgun/Mailgun.php [hunk 8]
```diff
@@ -208,4 +279,28 @@ public function BatchMessage($workingDomain, $autoSend = true)
...
+          return null;
+        }
+
+        $region = $this->regionForUrl($messageUrl);
+        $regionClient = $this->regionClients[$region];
+        $query = substr(
+            $messageUrl,
+            strspn(
+                $messageUrl,
+                'https://' . $this->regions[$region] . '/' . $regionClient->getApiVersion() . '/'
```
Comment (Nyholm) [inline src/Mailgun/Mailgun.php]: Wouldn't `'https://' . $this->regions[$region] . '/' . $regionClient->getApiVersion() . '/'` be equal to the first part of `$messageUrl` at this point?

**Comment by Nyholm** [comment]
I do also miss unit tests and maybe function tests for this feature.

File: src/Mailgun/Mailgun.php (hunk 4 above)
Comment (pirogoeth) [inline src/Mailgun/Mailgun.php]: Removed.

File: src/Mailgun/Mailgun.php (hunk 6 above)
Comment (pirogoeth) [inline src/Mailgun/Mailgun.php]: Wow, missed that. Thanks!

File: src/Mailgun/Mailgun.php (hunk 7 above)
Comment (pirogoeth) [inline src/Mailgun/Mailgun.php]: Good point. Fixed.

File: src/Mailgun/Connection/RestClient.php [hunk 9]
```diff
@@ -349,6 +349,14 @@ public function setApiVersion($apiVersion)
     }
 
     /**
+     * @return string
+     */
+    public function getApiVersion()
```
Comment (pirogoeth) [inline src/Mailgun/Connection/RestClient.php]: Let's talk about this tomorrow.

File: src/Mailgun/Mailgun.php (hunk 1 above)
Comment (pirogoeth) [inline src/Mailgun/Mailgun.php]: Fixed.

File: src/Mailgun/Mailgun.php (hunk 3 above)
Comment (pirogoeth) [inline src/Mailgun/Mailgun.php]: Fixed.

File: src/Mailgun/Mailgun.php (hunk 2 above)
Comment (pirogoeth) [inline src/Mailgun/Mailgun.php]: Fixed.

**Comment by pirogoeth** [comment]
@Nyholm Sorry about that, it's been a bit since I've used PHP. I'm gonna make sure everything gets fixed up and I'll have tests done either later today or tomorrow.

File: src/Mailgun/Mailgun.php (hunk 8 above)
Comment (pirogoeth) [inline src/Mailgun/Mailgun.php]: Working on making this less crummy.

File: src/Mailgun/Connection/RestClient.php (hunk 9 above)
Comment (DavidGarciaCat) [inline src/Mailgun/Connection/RestClient.php]: From my POV I cannot see the problem to implement a getter. Given we already have a setter, a getter is not a bad idea. @Nyholm why do you think this is not a good idea?

File: src/Mailgun/Mailgun.php [hunk 10]
```diff
@@ -166,6 +224,14 @@ public function setApiVersion($apiVersion)
     {
         $this->restClient->setApiVersion($apiVersion);
 
+        foreach ($this->regionClients as $regionClient) {
+            if ($regionClient == null) {
```
Comment (DavidGarciaCat) [inline src/Mailgun/Mailgun.php]: This should be compared strictly `===` or using the PHP function `is_null()`.

Also, please consider to use Yoda conditions, specially when you compare using `==`. Example:
`if ($regionClient = null)` will return **always** true and this is a human error but `if (null = $regionClient)` will return an error and it means we will prevent an unexpected behaviour.

File: src/Mailgun/Mailgun.php [hunk 11]
```diff
@@ -208,4 +274,29 @@ public function BatchMessage($workingDomain, $autoSend = true)
...
+
+    /**
+     * @param string $messageUrl
+     *
+     * @return \stdClass
+     */
+    public function getMessageFromUrl($messageUrl)
+    {
+        if ($messageUrl === null) {
+            return null;
```
Comment (DavidGarciaCat) [inline src/Mailgun/Mailgun.php]: According to PSR standards, when we `return null;` we can just write `return;`

Also, when we return `null` is must be declared on `@return` annotation too, so `@return \stdClass|null`

File: src/Mailgun/Connection/RestClient.php (hunk 9 above)
Comment (Nyholm) [inline src/Mailgun/Connection/RestClient.php]: I would be happy to remove the setter as well =)

I have an idea that the RestClient should only care about building a request and handle the response. That includes adding headers and exception handling. 

To increase flexibility of the RestClient we should not allow it to know about API version or the URLs. That should be handled in the MailGun class.

Comment (DavidGarciaCat) [inline src/Mailgun/Connection/RestClient.php]: That is a better approach, indeed. Good point.

File: src/Mailgun/Mailgun.php (hunk 11 above)
Comment (Nyholm) [inline src/Mailgun/Mailgun.php]: Agreed. Even though `void !== null`

Comment (pirogoeth) [inline src/Mailgun/Mailgun.php]: Added null to the return annotation and dropped null from the return statement.

File: src/Mailgun/Mailgun.php (hunk 10 above)
Comment (pirogoeth) [inline src/Mailgun/Mailgun.php]: Changed to strict comparison. I'll keep the Yoda conditions thing in mind, thanks! :)

File: src/Mailgun/Mailgun.php [hunk 12]
```diff
@@ -47,6 +60,51 @@ public function __construct(
...
+    {
+        foreach ($this->regions as $region => $apiUrl) {
+            $this->regionClients[$region] = new RestClient(
+                $this->apiKey,
+                $apiUrl,
+                $httpClient
+            );
+        }
+
+        return true;
```
Comment (DavidGarciaCat) [inline src/Mailgun/Mailgun.php]: ❓ Question: Why to add here `return true;`? If we are not going to return any other value does not make sense - from my POV - to return always `true` so we can just remove this. Anyway, if we throw an Exception (for example) we are not going to reach that `return` so IMHO does not make sense or is not adding a useful case.

File: src/Mailgun/Mailgun.php [hunk 13]
```diff
@@ -47,6 +60,51 @@ public function __construct(
...
+    private function regionForUrl($url)
+    {
+        foreach ($this->regions as $region => $apiUrl) {
+            if (strpos($url, 'https://' . $apiUrl) === 0) {
+                return $region;
+            }
+        }
+
+        // Region in URL does *not* exist in the regions mapping.
+        return null;
```
Comment (DavidGarciaCat) [inline src/Mailgun/Mailgun.php]: As discussed on my previous review, `return;` is the PSR standard for `return null;`

Also, I can see the `return @\string` annotation, but if I am not wrong `\string` is part of PHP7 annotations, isn't it? So I think a better approach here is `@return string` to be consistent with PHP5 code. In addition, given this method can return `null` too, it should be added to the annotation.

**Comment by pirogoeth** [comment]
Closing this since it was a little premature on my part.

**Pull request closed without merging**